### ISSUE TITLE
想请教一些训练细节

### ISSUE THREAD

**Issue by yjh0410** [thisisi3/Paddle-YOLOF#1]
在知乎搜到了您这个工作的文章，非常感兴趣！也非常感谢作者开源了这么优秀的YOLOF复现工作。
有一些关于训练和初始化的细节想请教一些作者：

1. **训练中的学习率策略**。官方YOLOF在设置学习率的时候是分了两部分，以batchsize=64、初始学习率为0.12为例，backbone部分的学习率则设置为0.04，即初始学习率的三分之一，其余部分的学习率则为0.12。在您的项目中，我看学习率设置为0.06，我想这是因为您将batchsize设置为32的缘故，那请问在您的项目中，backbone部分的学习率是否也设置为初始化学习率的三分之一，因为，我在您的源码中没有看到这个细节，所以，我想是不是您干脆就把整个网络的初始学习率都设置为了0.06？

2. **训练中的loss变成NAN的问题**。我发现官方的YOLOF会有这个问题，在MMDetection中也提到了这个训练不稳定的问题，请问您是否考虑到了这个问题，有使用哪些策略如梯度剪裁来避免这个问题吗？

3. **最后一个问题就是关于backbone了**。我看您使用的是caffe版的R50，我想请问一下，能否考虑使用torchvision版本的R-50。就我所知，这两版本的差别主要是在输入图像通道上，caffe版本用的BGR，torchvision版本用的RGB。另外，归一化方式好像也不太一样。这种差别的影响大吗？

以上是我感兴趣的几个问题，其中问题1和2是我最关心的。还望作者百忙之余，拨冗答疑，非常感谢，感激不尽！

**Comment by yjh0410** [comment]
@thisisi3 不好意思，还有一个感兴趣的问题就是您在VOC2007 test上的AP50是多少？然后才开始在COCO上去train的。

**Comment by thisisi3** [comment]
感谢关注！
1，训练中的学习率策略：Paddle Detection下面的ResNet支持自定义lr，所以可以通过修改config来实现，具体参数是lr_mult_list。
2，训练中的loss变成NAN的问题：这个问题在一开始我就关注了，因为MMDetection中有提到。我的版本中也会出现这个问题，但是一直没有找到解决方案。
3，最后一个问题就是关于backbone了：我是用caffe版本主要是为了和官方保持一致，你如果想尝试torchvision的版本，确实你提到的两点是需要注意的。我觉得两者的差别应该不会很大，但是我个人不建议使用其他版本的backbone，因为这样就不好对比基线。
4，VOC2007的结果：6次训练中会有一次loss不正常，其他5次大概是mean=45.6，std=0.26。我这里指的是COCO标准下的mAP，不是VOC下面的mAP，请知晓。

希望对你有帮助，有什么问题可以尽管提出，我会尽量回答，我也在试图改进这个项目，所以我们之间的讨论也是对我很有帮助的。

**Comment by yjh0410** [comment]
@thisisi3 收到，非常感谢作者的答复！
1. 我在您的项目中找到了lr_multi_list的设置，解决了我的疑惑。
2. 了解，那看来确实是一个待解决的问题。
3. 好的，因为我平时用惯了torchvision。遵从您的建议，我尝试一下caffe版的R-50。

**Comment by yjh0410** [comment]
@thisisi3 
您好，我最近歪打正着地找到了YOLOF为什么出现NAN的问题了，主要是下面这句话：

normalized_cls_logits = cls_logits + objectness - paddle.log(
                1.0 + paddle.clip(cls_logits.exp(), max=INF) +
                paddle.clip(objectness.exp(), max=INF))

这段代码中有个exp操作，并用clip剪裁一下，避免爆炸，但这个还是有隐患，就是在剪裁之前，exp可能已经爆炸溢出了，这个时候clip就没用了。所以，我改成先clip再exp，假定INF=1e8:

normalized_cls_pred = cls_pred + obj_pred - torch.log(
                1. + 
                torch.clamp(cls_pred, max=DEFAULT_EXP_CLAMP).exp() + 
                torch.clamp(obj_pred, max=DEFAULT_EXP_CLAMP).exp())

上面是torch的代码，其中DEFAULT_EXP_CLAMP = log(INF)。这样修改后，就没再遇到NAN的问题了。

**Comment by thisisi3** [comment]
@yjh0410
Great work! Thanks.
I will refer your findings in another issue, hope you do not mind.

**Comment by yjh0410** [comment]
@thisisi3 

You are welcome. To be honest, I'm not 100% sure that this is the main reason for NAN, but from my personal experience, it is true that when I modified this code, the NAN problem magically disappeared. Hope you don't run into NAN problems again on your excellent YOLOF project.

**Comment by thisisi3** [comment]
I think this is very likely to be the reason. Even if it is not, it is still a good practice to clip value before exp() operation.
